feat: add C implementation for `stats/base/ndarray/dmidrange` (#10077)
stdlib-js/stdlib: lib/node_modules/@stdlib/stats/base/ndarray/dmidrange/README.md

@license Apache-2.0

> Copyright (c) 2026 The Stdlib Authors.

Reviewer: Don't update the copyright of existing files.

> Copyright (c) 2026 The Stdlib Authors.

Reviewer: ```suggestion
Copyright (c) 2025 The Stdlib Authors.
```

<!-- /.examples -->

<!-- C interface documentation. -->

* * *

<section class="c">

## C APIs

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- C usage documentation. -->

<section class="usage">

### Usage

```c
#include "stdlib/stats/base/ndarray/dmidrange.h"
```

#### stdlib_stats_dmidrange( arrays )

> Computes the midrange of a one-dimensional double-precision floating-point ndarray.

Reviewer: ```suggestion
Computes the mid-range of a one-dimensional double-precision floating-point ndarray.
```

```c
#include "stdlib/ndarray/ctor.h"
#include "stdlib/ndarray/dtypes.h"
#include "stdlib/ndarray/index_modes.h"
#include "stdlib/ndarray/orders.h"
#include "stdlib/ndarray/base/bytes_per_element.h"
#include <stdint.h>

// Create an ndarray:
const double data[] = { 1.0, 2.0, 3.0, 4.0 };
int64_t shape[] = { 4 };
int64_t strides[] = { STDLIB_NDARRAY_FLOAT64_BYTES_PER_ELEMENT };
int8_t submodes[] = { STDLIB_NDARRAY_INDEX_ERROR };

struct ndarray *x = stdlib_ndarray_allocate( STDLIB_NDARRAY_FLOAT64, (uint8_t *)data, 1, shape, strides, 0, STDLIB_NDARRAY_ROW_MAJOR, STDLIB_NDARRAY_INDEX_ERROR, 1, submodes );

> // Compute the midrange:

Reviewer: ```suggestion
// Compute the mid-range:
```

Be consistent in your spelling. We use `mid-range` for JS; we should use `mid-range` for C.

    // Specify the number of array dimensions:
    const int64_t ndims = 1;

    // Specify the array shape:
    int64_t shape[] = { 4 };

    // Specify the array strides:
    int64_t strides[] = { 2*STDLIB_NDARRAY_FLOAT64_BYTES_PER_ELEMENT };

    // Specify the byte offset:
    const int64_t offset = 0;

    // Specify the array order:
    const enum STDLIB_NDARRAY_ORDER order = STDLIB_NDARRAY_ROW_MAJOR;

    // Specify the index mode:
    const enum STDLIB_NDARRAY_INDEX_MODE imode = STDLIB_NDARRAY_INDEX_ERROR;

    // Specify the subscript index modes:
    int8_t submodes[] = { STDLIB_NDARRAY_INDEX_ERROR };
    const int64_t nsubmodes = 1;

    // Create an ndarray:
    struct ndarray *x = stdlib_ndarray_allocate( STDLIB_NDARRAY_FLOAT64, (uint8_t *)data, ndims, shape, strides, offset, order, imode, nsubmodes, submodes );
    if ( x == NULL ) {
        fprintf( stderr, "Error allocating memory.\n" );
        exit( 1 );
    }

    // Define a list of ndarrays:
    const struct ndarray *arrays[] = { x };

    // Compute the midrange:
    double v = stdlib_stats_dmidrange( arrays );

    // Print the result:

>     printf( "midrange: %lf\n", v );

Reviewer: ```suggestion
    // Compute the mid-range:
    double v = stdlib_stats_dmidrange( arrays );

    // Print the result:
    printf( "mid-range: %lf\n", v );
```

<!-- /.examples -->

<!-- C interface documentation. -->

* * *

<section class="c">

## C APIs

<!-- Section to include introductory text. Make sure to keep an empty line after the intro `section` element and another before the `/section` close. -->

<section class="intro">

</section>

<!-- /.intro -->

<!-- C usage documentation. -->

<section class="usage">

### Usage

```c
#include "stdlib/stats/base/ndarray/dmidrange.h"
```

#### stdlib_stats_dmidrange( arrays )

> Computes the mid-range of a one-dimensional double-precision floating-point ndarray.

Reviewer: ```suggestion
Computes the [mid-range][mid-range] of a one-dimensional double-precision floating-point ndarray.
```

    const int64_t ndims = 1;

    // Specify the array shape:
    int64_t shape[] = { 4 };

    // Specify the array strides:
    int64_t strides[] = { 2*STDLIB_NDARRAY_FLOAT64_BYTES_PER_ELEMENT };

    // Specify the byte offset:
    const int64_t offset = 0;

    // Specify the array order:
    const enum STDLIB_NDARRAY_ORDER order = STDLIB_NDARRAY_ROW_MAJOR;

    // Specify the index mode:
    const enum STDLIB_NDARRAY_INDEX_MODE imode = STDLIB_NDARRAY_INDEX_ERROR;

    // Specify the subscript index modes:
    int8_t submodes[] = { STDLIB_NDARRAY_INDEX_ERROR };
    const int64_t nsubmodes = 1;

    // Create an ndarray:
    struct ndarray *x = stdlib_ndarray_allocate( STDLIB_NDARRAY_FLOAT64, (uint8_t *)data, ndims, shape, strides, offset, order, imode, nsubmodes, submodes );
    if ( x == NULL ) {
        fprintf( stderr, "Error allocating memory.\n" );
        exit( 1 );
    }

    // Define a list of ndarrays:
    const struct ndarray *arrays[] = { x };

    // Compute the midrange:
    double v = stdlib_stats_dmidrange( arrays );

    // Print the result:
    // Compute the mid-range:
    double v = stdlib_stats_dmidrange( arrays );

    // Print the result:

>     printf( "mid-range: %lf\n", v );

Reviewer: ```suggestion
    // Compute the mid-range:
    double v = stdlib_stats_dmidrange( arrays );

    // Print the result:
    printf( "mid-range: %lf\n", v );
```
